src/lib.rs

```diff
@@ -803,6 +803,74 @@ impl<'a> VariantInfo<'a> {
         self
     }
 
+    /// Consumes this `Variant` object, creating two new `Variant` objects from it.
+    ///
+    /// The predicate passed to partition() can return true, or false. partition() returns a pair,
+    /// an object with the bindings for which it returned true, and another object with the
+    /// bindings for which it returned false.
+    ///
+    /// # Example
+    /// ```
+    /// # use synstructure::*;
+    /// let di: syn::DeriveInput = syn::parse_quote! {
+    ///     enum A {
+    ///         B{ a: i32, b: i32 },
+    ///         C{ a: u32 },
+    ///     }
+    /// };
+    /// let mut s = Structure::new(&di);
+    ///
+    /// let (a, b) = s.variants()[0].clone().partition(|bi| {
+    ///     bi.ast().ident == Some(quote::format_ident!("a"))
+    /// });
+    ///
+    /// assert_eq!(
+    ///     a.each(|bi| quote!(println!("{:?}", #bi))).to_string(),
+    ///
+    ///     quote!{
+    ///         A::B{ a: ref __binding_0, .. } => {
+    ///             { println!("{:?}", __binding_0) }
+    ///         }
+    ///     }.to_string()
+    /// );
+    ///
+    /// assert_eq!(
+    ///     b.each(|bi| quote!(println!("{:?}", #bi))).to_string(),
+    ///
+    ///     quote!{
+    ///         A::B{ b: ref __binding_1, .. } => {
+    ///             { println!("{:?}", __binding_1) }
+    ///         }
+    ///     }.to_string()
+    /// );
+    /// ```
+    pub fn partition<F>(mut self, mut f: F) -> (Self, Self)
```

Comment:
I wonder a bit if `partition` is the behaviour we'd want here, or if perhaps the desirable behaviour would actually be something closer to `drain_filter`, so that it mutates `self` in-place, and then returns a second `Variant` which has only the filtered-out bindings. I think an API like that might be more generally useful, and could be used to implement partition if a downstream crate actually wanted it. 

Could we look into switching this PR to implement a `drain_filter`-like API instead?

Reply:
I think `drain_filter` would be an improvement to avoid mixing which element of the tuple has what so yeah I'll change it. Does this seem all-right?
```rust
pub fn drain_filter<F(&mut self, mut f: F) -> Self where F: FnMut(&BindingInfo<'_>) -> bool,
```